### PR TITLE
Start the simulator stopped.

### PR DESCRIPTION
**Pull request by microbit-matt-hillsdon**
- "flash" starts the simulator
- "stop" can now stop it entirely, rather than drop to the REPL
- "restart" restarts the current program.

### CODE REVIEW

File: src/board/ui.ts [hunk 1]
```diff
@@ -63,12 +72,47 @@ export class BoardUI {
...
     this.buttons.forEach((b) => b.initialize());
     this.pins.forEach((p) => p.initialize());
     this.display.initialize();
     this.accelerometer.initialize();
+    this.serialInputBuffer = [];
+  }
+
+  async stop(): Promise<void> {
+    await this.operations.stop();
+    window.board.serialInputBuffer.push(3, 4);
```
Comment (microbit-matt-hillsdon) [inline src/board/ui.ts:93]: Let's reinstate the comment this used to have.

Comment (microbit-matt-hillsdon) [inline src/board/ui.ts:93]:
```suggestion
    // Ctrl-C to interrupt, Ctrl-D to restart but requesting the stop prevents actual restart.
    window.board.serialInputBuffer.push(3, 4);
```

File: src/board/listener.ts [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+    if (this.stoppedPromise) {
+      throw new Error("Already started!");
+    }
+    this.stoppedPromise = this.main!();
+  }
+
+  async stop(): Promise<void> {
+    if (this.stoppedPromise) {
+      this._requestStop!();
+      window.board.serialInputBuffer.push(3, 4);
```
Comment (microbit-matt-hillsdon) [inline src/board/listener.ts]: This is really dubious layering.

File: src/board/ui.ts (hunk 1 above)
Comment (microbit-matt-hillsdon) [inline src/board/ui.ts]: This is madness. We are "board" and this is also happening inside operations.stop() where it is correctly sequenced.

File: src/board/ui.ts [hunk 3]
```diff
@@ -63,12 +72,47 @@ export class BoardUI {
...
+  }
+
+  async flash(filesystem: Record<string, Uint8Array>): Promise<void> {
+    await this.operations.stop();
+    this.fs.clear();
+    Object.entries(filesystem).forEach(([name, value]) => {
+      const idx = this.fs.create(name);
+      this.fs.write(idx, value);
+    });
+    console.log(this.fs);
```
Comment (microbit-matt-hillsdon) [inline src/board/ui.ts]: Remove debug

```suggestion
```

**Comment by microbit-matt-hillsdon** [comment]
This seems to work as expected but it's unexpectedly slow to run a new program. Not yet looked at what step is slow (suspect awaiting the main C entry point).

**Comment by microbit-matt-hillsdon** [comment]
> This seems to work as expected but it's unexpectedly slow to run a new program. Not yet looked at what step is slow (suspect awaiting the main C entry point).

@microbit-robert and I can no longer reproduce this. At some point I had a hacky local version with a 3 second timeout, perhaps I confused myself and used an old browser tab.

**Review by microbit-robert — APPROVED** [review]
LGTM.

**Pull request merged**